contracts/SolverTrampoline.sol

```diff
@@ -17,17 +23,41 @@ contract SolverTrampoline {
     /// @dev The domain separator for signing EIP-712 settlements.
     bytes32 public immutable domainSeparator;
 
+    /// @dev Nonce by solver address.
+    mapping(address => uint256) public nonces;
+
     constructor(Settlement settlementContract_) {
         settlementContract = settlementContract_;
         solverAuthenticator = settlementContract_.authenticator();
-        
+
         domainSeparator = keccak256(abi.encode(
             keccak256("EIP712Domain(uint256 chainId,address verifyingContract)"),
             block.chainid,
             address(this)
         ));
     }
 
+    /// @dev Executes a settlement on behalf of a solver.
+    function settle(bytes calldata settlement, uint256 nonce, bytes32 r, bytes32 s, uint8 v) external {
+        bytes32 messageDigest = settlementMessage(settlement, nonce);
+        address solver = ecrecover(messageDigest, v, r, s);
+        if (solver == address(0) || !solverAuthenticator.isSolver(solver)) {
+            revert Unauthorized();
+        }
+        if (nonce != nonces[solver]) {
+            revert InvalidNonce();
+        }
+
+        nonces[solver] = nonce + 1;
+        (bool success, bytes memory data) = address(settlementContract).call(settlement);
+        if (!success) {
+            // Propagate the revert error.
+            assembly {
+                revert(add(data, 32), mload(data))
+            }
```

Comment:
Why all of this dark magic, instead of just calling `settlementContract.settle(settlement)` (please consider adding a comment)?

Reply:
This is so we don't need to encode the `settle` function which would have been A LOT more gas. Also, this allows for a few more things when calling the settlement contract:
- To use the "fast path" that we never used (since we aren't calling the specific method anymore)
- To issue cancellations (the settlement contract has a fallback function, so calling it without calldata is allowed, enabling `settle("0x", nonce, r, s, v)` to be executed by solvers directly as a cancellation mechanism).

Reply:
Also, conceptually, doing a low-level `.call()` means the trampoline contract has no knowledge/understanding of what is actually being called on the settlement contract.